clusterloader2/pkg/measurement/common/network/controller.go

```diff
@@ -0,0 +1,438 @@
+		if firstClientPodTime == 0 {
+			firstClientPodTime = getTimeStampForPod()
+		}
+		sendReqToClient(uniqPodPair, protocol, duration, firstClientPodTime)
+	}
+
+}
+
+func formUniquePodPair(originalMap map[string][]workerPodData) {
+	var uniqPodPair uniquePodPair
+	lastPodPair := uniquePodPair{IsLastPodPair: true}
+
+	var i = 0
+
+	for {
+		for key, value := range originalMap {
+			unUsedPod, err := getUnusedPod(&value)
+			originalMap[key] = value
+			if err != nil {
+				delete(originalMap, key)
+				continue
+			}
+			i++
+
+			if i == 1 {
+				uniqPodPair.SrcPodIp = unUsedPod.podIp
+				uniqPodPair.SrcPodName = unUsedPod.podName
+			} else if i == 2 {
+				uniqPodPair.DestPodIp = unUsedPod.podIp
+				uniqPodPair.DestPodName = unUsedPod.podName
+				i = 0
+				uniqPodPairList = append(uniqPodPairList, uniqPodPair)
+				podPairCh <- uniqPodPair
+			}
+		}
+		if len(originalMap) == 0 {
+			podPairCh <- lastPodPair
+			break
+		}
+	}
+}
+
+func getUnusedPod(unusedPodList *[]workerPodData) (workerPodData, error) {
+	var unusedPod workerPodData
+	if len(*unusedPodList) == 0 {
+		return unusedPod, errors.New("unused pod list empty")
+	}
+	numOfPods := len(*unusedPodList)
+	//extract last pod of slice
+	unusedPod = (*unusedPodList)[numOfPods-1]
+	*unusedPodList = (*unusedPodList)[:numOfPods-1]
+	return unusedPod, nil
+}
+
+func sendReqToClient(uniqPodPair uniquePodPair, protocol string, duration int, futureTimestamp int64) {
+	klog.V(3).Infof("sending req to cli-pod: %s, protocol: %s, futureTimeStamp: %v, duration: %v",
+		uniqPodPair.SrcPodName, protocol, futureTimestamp, duration)
+	switch protocol {
+	case Protocol_TCP:
+		startWork(uniqPodPair.SrcPodName, httpPathMap[TCP_Client], duration, futureTimestamp, "", uniqPodPair.DestPodIp)
+	case Protocol_UDP:
+		startWork(uniqPodPair.SrcPodName, httpPathMap[UDP_Client], duration, futureTimestamp, "", uniqPodPair.DestPodIp)
+	case Protocol_HTTP:
+		startWork(uniqPodPair.SrcPodName, httpPathMap[HTTP_Client], duration, futureTimestamp, "", uniqPodPair.DestPodIp)
+	}
+}
+
+func sendReqToSrv(uniqPodPair uniquePodPair, protocol string, duration int) {
+	klog.V(3).Infof("sending req to srv-pod: %s, protocol: %s", uniqPodPair.DestPodName, protocol)
+	switch protocol {
+	case Protocol_TCP:
+		startWork(uniqPodPair.DestPodName, httpPathMap[TCP_Server], duration, time.Now().Unix(), "1", "")
+	case Protocol_UDP:
+		startWork(uniqPodPair.DestPodName, httpPathMap[UDP_Server], duration, time.Now().Unix(), "1", "")
+	case Protocol_HTTP:
+		startWork(uniqPodPair.DestPodName, httpPathMap[HTTP_Server], duration, time.Now().Unix(), "1", "")
+	}
+}
+
+func getTimeStampForPod() int64 {
+	currTime := time.Now()
+	initDelay := time.Second * time.Duration(initialDelayForTCExec)
+	futureTime := currTime.Add(initDelay).Unix()
+	return futureTime
+}
+
+func collectMetrics(uniqPodPair uniquePodPair, protocol string) *MetricResponse {
+	var podName string
+
+	switch protocol {
+	case Protocol_TCP:
+		fallthrough
```

Comment:
Don't use fallthrough

should be

switch protocol {
case ProtocolTCP, ProtocolUDP:
  // something
case ProtocolHTTP:
  // something
}

Reply:
modified the code as per comments.